File: swift/Utils/ZipUtils.m
#import "ZipUtils.h"

@implementation ZipUtils

-(void) unzip {
    [NSThread detachNewThreadSelector:@selector(doUncompress) toTarget:self withObject:nil];
    
}

-(void) doUncompress {
    if ([self.delegate respondsToSelector:@selector(zipWillUnzip)]) {
        if ([self.delegate zipWillUnzip]) {
            ZipArchive * za = [[ZipArchive alloc] init];
            NSNumber * ret = [NSNumber numberWithBool:NO];
            if ([za UnzipOpenFile:self.archiveFile]) {
                BOOL succ = [za UnzipFileTo:self.extractPath overWrite:YES];
                [za UnzipCloseFile];
                ret = [NSNumber numberWithBool:succ];
            }
            [self performSelectorOnMainThread:@selector(callback:) withObject:ret waitUntilDone:YES];
        }
    }
}

-(void)callback: (NSNumber *) ret {
    if ([self.delegate respondsToSelector:@selector(ziputils:unzipCompleted:)]) {
        [self.delegate ziputils:self unzipCompleted:[ret boolValue]];
    }
}

@end

File: swift/Utils/ZipUtils.swift
import UIKit

@objc protocol ZipUtilsDelegate: NSObjectProtocol {
    @objc optional func zipWillUnzip() -> Bool
    @objc optional func ziputils(ziputils: ZipUtils, unzipCompleted succ: Bool)

}

class ZipUtils: NSObject {
    var delegate: ZipUtilsDelegate?
    var archiveFile: String?
    var extractPath: String?
    
    func unzip() {
        Thread.detachNewThreadSelector(#selector(doUncompress), toTarget: self, with: nil)
    }
    
    func doUncompress() {
        var b = false
        if (self.delegate?.zipWillUnzip != nil) {
            b = self.delegate!.zipWillUnzip!()
        }
        if (b) {
            let za = ZipArchive()
            var ret = NSNumber(value: false)
            if (za.unzipOpenFile(self.archiveFile!)) {
                let succ = za.unzipFile(to: self.extractPath, overWrite:true)
                za.unzipCloseFile()
                ret = NSNumber(value: succ)
            }
            DispatchQueue.main.async(execute: {
                self.callback(ret: ret)
            })
        }
    }
    
    func callback(ret: NSNumber) {
        self.delegate?.ziputils?(ziputils: self, unzipCompleted: ret.boolValue)
    }

}
